File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/highriskvenuesupload/model/RiskyWindow.java
package uk.nhs.nhsx.highriskvenuesupload.model;

import java.util.Objects;

public class RiskyWindow {
    public final String from;
    public String until;

    public RiskyWindow(String from, String until) {
        this.from = from;
        this.until = until;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        RiskyWindow that = (RiskyWindow) o;
        return from.equals(that.from) &&
                until.equals(that.until);
    }

    @Override
    public int hashCode() {
        return Objects.hash(from, until);
    }
}

File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/highriskvenuesupload/model/RiskyWindow.kt
package uk.nhs.nhsx.highriskvenuesupload.model

import java.time.Instant
import java.time.format.DateTimeParseException

data class RiskyWindow(val from: Instant, val until: Instant) {

    init {
        require(from.isBefore(until)) {
            "validation error: Start date must be <= end date"
        }
    }

    companion object {
        fun of(from: String, until: String): RiskyWindow = try {
            RiskyWindow(Instant.parse(from), Instant.parse(until))
        } catch (e: DateTimeParseException) {
            throw IllegalArgumentException("validation error: Date did not conform to expected format yyyy-MM-dd'T'HH:mm:ss'Z'")
        }
    }
}
